src/main/java/org/opensearch/security/resources/api/share/ShareRequest.java

```diff
@@ -0,0 +1,219 @@
+    private final String resourceId;
+    @JsonProperty("resource_index")
+    private final String resourceIndex;
+    @JsonProperty("share_with")
+    private final ShareWith shareWith;
+    @JsonProperty("patch")
+    private final Map<String, Object> patch;
+
+    private final RestRequest.Method method;
+
+    /**
+     * Private constructor to enforce usage of Builder
+     */
+    private ShareRequest(Builder builder) {
+        this.resourceId = builder.resourceId;
+        this.resourceIndex = builder.resourceIndex;
+        this.shareWith = builder.shareWith;
+        this.patch = builder.patch;
+        this.method = builder.method;
+    }
+
+    /**
+     * Static factory method to initialize ShareRequest from a Map.
+     */
+    @SuppressWarnings("unchecked")
+    public static ShareRequest from(Map<String, Object> source) throws IOException {
+        Builder builder = new Builder();
+
+        builder.method((RestRequest.Method) source.get("method"));
+
+        builder.resourceId((String) source.get("resource_id"));
+
+        builder.resourceIndex((String) source.get("resource_index"));
+
+        if (source.containsKey("share_with")) {
+            builder.shareWith((Map<String, Object>) source.get("share_with"));
+        }
+
+        if (source.containsKey("patch")) {
+            builder.patch((Map<String, Object>) source.get("patch"));
+        }
+
+        return builder.build();
+    }
+
+    public ShareRequest(StreamInput in) throws IOException {
+        super(in);
+        this.resourceId = in.readString();
+        this.resourceIndex = in.readString();
+        this.shareWith = in.readOptionalWriteable(ShareWith::new);
+        this.patch = in.readMap();
+        this.method = in.readEnum(RestRequest.Method.class);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(resourceId);
+        out.writeString(resourceIndex);
+        if (shareWith != null) {
+            shareWith.writeTo(out);
+        }
+        if (patch != null) {
+            out.writeString(patch.toString());
+        }
+        out.writeEnum(method);
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        var arv = new ActionRequestValidationException();
+        if (Strings.isNullOrEmpty(resourceIndex) || Strings.isNullOrEmpty(resourceId)) {
+            arv.addValidationError("resource_id and resource_index must be present");
+            throw arv;
+        }
+
+        // no further check needed in case of GET
+        if (method == RestRequest.Method.GET) {
+            return null;
+        }
+        // either of shareWith or patch must be present in the request
+        if (shareWith == null && (patch == null || patch.isEmpty())) {
+            String message = method == RestRequest.Method.PATCH ? "Patch" : "ShareWith";
+            arv.addValidationError(message + " is required");
+            throw arv;
+        }
```

Comment:
if `patch` is present do you then also verify that the map contains `add` and `revoke` somewhere?

Reply:
yes, I added a validation that either add or revoke must be present.